### PR TITLE
FFT invokers

### PR DESCRIPTION
**Pull request by DrizztDoUrden**

### CODE REVIEW

**Comment by DrizztDoUrden** [comment]
Tbh I don't understand the state of this branch on Jenkins

**Comment by DrizztDoUrden** [comment]
I don't get it, Jenkins shows it as passed.

**Comment by atamazov** [comment]
@DrizztDoUrden Don't mind. There are issues in Jenkins <--> github communication. Your PR has passed CI testing.

**Comment by atamazov** [comment]
I am beginning the review/testing process.

**Comment by atamazov** [comment]
Config:
```
./bin/MIOpenDriver conv -x 5 -y 5 -W 28 -H 28 -c 192 -n 16 -k 32 -p 2 -q 2 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0 -V 1 -s 0 -w 2 -t 1 -i 6
```
Baseline's log:
```
...
MIOpen(HIP): Info [FindConvFwdAlgorithm] requestAlgoCount = 2, workspace = 81330176
...
MIOpen(HIP): Info [SetValues] 192-28-28-5x5-32-28-28-16-2x2-1x1-1x1-0-NCHW-FP32-F, content inserted: miopenConvolutionFwdAlgoFFT:fft,0.466878,81330176,miopenConvolutionFwdAlgoFFT,192x28x28x5x5x32x28x28x16xNCHWxFP32x2x2x1x1x1x1x1xF
MIOpen(HIP): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoWinograd	0.328797	0
MIOpen(HIP): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoFFT	0.466878	81330176
MIOpen(HIP): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoImplicitGEMM	1.03295	0
MIOpen(HIP): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoDirect	1.23791	0
MIOpen(HIP): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoGEMM	2.72382	15052800
MIOpen(HIP): Info [FindConvFwdAlgorithm] FW Chosen Algorithm: ConvBinWinogradRxSf2x3 , 0, 0.328797
...
```
PR's log:
```
...
MIOpen(HIP): Info [FindConvFwdAlgorithm] requestAlgoCount = 2, workspace = 15052800
...
MIOpen(HIP): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoWinograd	0.327839	0
MIOpen(HIP): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoImplicitGEMM	1.01696	0
MIOpen(HIP): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoDirect	1.24336	0
MIOpen(HIP): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoGEMM	2.69056	15052800
MIOpen(HIP): Info [FindConvFwdAlgorithm] FW Chosen Algorithm: ConvBinWinogradRxSf2x3 , 0, 0.327839
...
```
As you can see, FFT records are missing from the log. Most likely, a problem with applicability.

**Review by atamazov — CHANGES_REQUESTED** [review]
The problem shown at https://github.com/ROCmSoftwarePlatform/MIOpen/pull/455#issuecomment-710346306 should be addressed.

**Comment by DrizztDoUrden** [comment]
@atamazov Updated

File: src/solver/fft.cpp [hunk 1]
```diff
@@ -102,34 +107,91 @@ static void cgemm_grid(size_t* global_work_size,
...
     std::tie(in_n, in_c, in_h, in_w) = miopen::tien<4>(xDesc.GetLengths());
 
-    (void)wDesc;
+    int out_n, out_c, out_h, out_w;
+    std::tie(out_n, out_c, out_h, out_w) = miopen::tien<4>(yDesc.GetLengths());
+
+    int wei_k, wei_c, wei_h, wei_w;
+    std::tie(wei_k, wei_c, wei_h, wei_w) = miopen::tien<4>(wDesc.GetLengths());
+
+    bool supported = true;
```
Comment (atamazov) [inline src/solver/fft.cpp:135]: This block of code looks ugly. Suggestion:
```C++
if(in_n < 1 || in_n > 512 || wei_k < 1 || wei_k > 512 || in_c * in_n % 16 != 0 ||
   wei_c * wei_k % 16 != 0 || out_c * out_n % 16 != 0)
    return false;

if(std::tie(in_h, in_w) != std::make_tuple(28, 28) &&
   std::tie(in_h, in_w) != std::make_tuple(27, 27) &&
   std::tie(in_h, in_w) != std::make_tuple(14, 14) &&
   std::tie(in_h, in_w) != std::make_tuple(7, 7))
    return false;   
```

File: src/solver/fft.cpp [hunk 2]
```diff
@@ -102,34 +107,91 @@ static void cgemm_grid(size_t* global_work_size,
...
+    const auto is_fwd    = ctx.direction.IsForward();
+    decltype(auto) conv  = ctx.conv_problem.GetConv();
+    decltype(auto) xDesc = is_fwd ? ctx.conv_problem.GetIn() : ctx.conv_problem.GetOut();
+    decltype(auto) yDesc = is_fwd ? ctx.conv_problem.GetOut() : ctx.conv_problem.GetIn();
+    decltype(auto) wDesc = ctx.conv_problem.GetWeights();
 
     if(xDesc.GetType() != miopenFloat || wDesc.GetType() != miopenFloat ||
        yDesc.GetType() != miopenFloat)
-        return -1;
+        return false;
```
Comment (atamazov) [inline src/solver/fft.cpp:124]:
```suggestion
if(!ctx.conv_problem.IsFp32())
    return false;
```

File: src/solver/fft.cpp [hunk 3]
```diff
@@ -102,34 +107,91 @@ static void cgemm_grid(size_t* global_work_size,
...
+                    : supported;
+
+    const auto cparam = std::make_tuple(conv.GetConvPads()[0],
+                                        conv.GetConvPads()[1],
+                                        conv.GetConvStrides()[0],
+                                        conv.GetConvStrides()[1]);
+
+    supported = (std::tie(wei_h, wei_w) != std::make_tuple(5, 5)) ? false : supported;
+    supported = (cparam != std::make_tuple(2, 2, 1, 1)) ? false : supported;
+    supported = (yDesc.GetType() != miopenFloat) ? false : supported;
```
Comment (atamazov) [inline src/solver/fft.cpp:160]: This check is not necessary

File: src/solver/fft.cpp [hunk 4]
```diff
@@ -102,34 +107,91 @@ static void cgemm_grid(size_t* global_work_size,
...
+
+    return supported;
```
Comment (atamazov) [inline src/solver/fft.cpp:162]:
```suggestion
return std::tie(wei_h, wei_w) == std::make_tuple(5, 5) &&
       cparam == std::make_tuple(2, 2, 1, 1);
```

**Review by atamazov — CHANGES_REQUESTED** [review]

File: src/solver/fft.cpp (hunk 1 above)
Comment (DrizztDoUrden) [reply src/solver/fft.cpp:135]: It is a copy-paste from convolutionocl.cpp or convolutionocl_fft.cpp or something like that. Would you like me to fix it?

File: src/solver/fft.cpp (hunk 2 above)
Comment (DrizztDoUrden) [reply src/solver/fft.cpp:124]: I'll fix that. By the way, you can make a multi-line suggestion by dragging the + image.

File: src/solver/fft.cpp (hunk 3 above)
Comment (DrizztDoUrden) [reply src/solver/fft.cpp:160]: It's been moved as is. Would you like me to fix it?

File: src/solver/fft.cpp [hunk 5]
```diff
@@ -102,34 +107,74 @@ static void cgemm_grid(size_t* global_work_size,
...
+bool FFT::IsApplicable(const ConvolutionContext& ctx) const
 {
-
-    if(workSpaceSize == 0)
-        return -1;
-
     // disable running any FFT based convolutions by checking this env variable
     if(miopen::IsDisabled(MIOPEN_DEBUG_CONV_FFT{}))
-        return -1;
+        return false;
```
[the remark this answers is not in the document]
Comment (DrizztDoUrden) [reply src/solver/fft.cpp:114]: Ok, will fix that.

**Comment by DrizztDoUrden** [comment]
Anyway resolved all the comments.

File: src/solver/fft.cpp (hunk 1 above)
Comment (atamazov) [reply src/solver/fft.cpp]: OMG I am sorry

File: src/solver/fft.cpp (hunk 3 above)
Comment (atamazov) [reply src/solver/fft.cpp]: Well... better fix.

**Comment by atamazov** [comment]
@DrizztDoUrden Thanks. The plus is that now no one will blame you for this code.

**Comment by atamazov** [comment]
These configs:
```
export MIOPEN_FIND_MODE=normal
./bin/MIOpenDriver conv -x 5 -y 5 -W 28 -H 28 -c 192 -n 16 -k 32 -p 2 -q 2 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0 -V 1 -s 0 -w 2 -t 1 -i 6 -C /spb-hlc-fs2/1/users/atamazov/miopen-verify2
./bin/MIOpenDriver conv -x 5 -y 5 -W 14 -H 14 -c 512 -n 16 -k 48 -p 2 -q 2 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0 -V 1 -s 0 -w 2 -t 1 -i 6 -C /spb-hlc-fs2/1/users/atamazov/miopen-verify2
./bin/MIOpenDriver conv -x 5 -y 5 -W 7 -H 7 -c 832 -n 16 -k 128 -p 2 -q 2 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0 -V 1 -s 0 -w 2 -t 1 -i 6 -C /spb-hlc-fs2/1/users/atamazov/miopen-verify2
```
fail with
```
MIOpen(HIP): Info [FindSolutionImpl] FFT (not searchable)
MIOpen Error: boost::bad_any_cast: failed conversion using boost::any_cast
```
Full logs:
- HIP backend: [smoke-fp32-93.sh.opencl.872f83e77.ddu_fft-invokers.0.fail.log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/5417211/smoke-fp32-93.sh.opencl.872f83e77.ddu_fft-invokers.0.fail.log.txt)
- OCL backend: [smoke-fp32-93.sh.hip-clang.872f83e77.ddu_fft-invokers.0.fail.log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/5417212/smoke-fp32-93.sh.hip-clang.872f83e77.ddu_fft-invokers.0.fail.log.txt)

File: src/solver/fft.cpp [hunk 6]
```diff
@@ -389,8 +389,8 @@ ConvSolution FFT::GetSolution(const ConvolutionContext& ctx) const
         int halfw = static_cast<int>(workSpaceSize) / (2 * 2 * static_cast<int>(sizeof(float)));
         const int padding = FFTConvParams::TransposePadding;
 
-        return [=](const Handle& handle, const boost::any& primitive_params) {
-            const auto params   = boost::any_cast<conv::DataInvokeParams>(primitive_params);
+        return [=](const Handle& handle, const AnyInvokeParams& primitive_params) {
+            const auto params   = primitive_params.CastTo<conv::DataInvokeParams>();
```
Comment (atamazov) [inline src/solver/fft.cpp:393]: @DrizztDoUrden clang-tidy error here

**Comment by atamazov** [comment]
@DrizztDoUrden <details><summary><b><i>The same configs are failing:</i></b></summary>

```
./bin/MIOpenDriver conv -x 5 -y 5 -W 28 -H 28 -c 192 -n 16 -k 32 -p 2 -q 2 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0 -V 1 -s 0 -w 2 -t 1 -i 6 -C /spb-hlc-fs2/1/users/atamazov/miopen-verify2
...
MIOpen(HIP): Info [FindConvFwdAlgorithm] requestAlgoCount = 2, workspace = 15052800
...
MIOpen(HIP): Info [FindSolutionImpl] FFT (not searchable)
MIOpen Error: /home/atamazov/github/MLOpen2/src/solver/fft.cpp:399: Not enough workspace for FFT: expected 81330176, got 15052800
RunForwardGPU() failed, rc = 0x7
Forward Convolution Failed: 1.79769e+308 > 1.5e-06
MIOpen(HIP): Info [FindConvBwdDataAlgorithm] requestAlgoCount = 2, workspace = 15052800
...
MIOpen(HIP): Info [FindSolutionImpl] FFT (not searchable)
MIOpen Error: /home/atamazov/github/MLOpen2/src/solver/fft.cpp:399: Not enough workspace for FFT: expected 81330176, got 15052800
MIOpen(HIP): Info [FindConvBwdWeightsAlgorithm] requestAlgoCount = 2, workspace = 71450624
...
./bin/MIOpenDriver conv -x 5 -y 5 -W 14 -H 14 -c 512 -n 16 -k 48 -p 2 -q 2 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0 -V 1 -s 0 -w 2 -t 1 -i 6 -C /spb-hlc-fs2/1/users/atamazov/miopen-verify2
...
MIOpen(HIP): Info [FindConvFwdAlgorithm] requestAlgoCount = 2, workspace = 10035200
...
MIOpen(HIP): Info [FindSolutionImpl] FFT (not searchable)
MIOpen Error: /home/atamazov/github/MLOpen2/src/solver/fft.cpp:399: Not enough workspace for FFT: expected 94740480, got 10035200
RunForwardGPU() failed, rc = 0x7
Forward Convolution Failed: 1.79769e+308 > 1.5e-06
MIOpen(HIP): Info [FindConvBwdDataAlgorithm] requestAlgoCount = 2, workspace = 10035200
...
MIOpen(HIP): Info [FindSolutionImpl] FFT (not searchable)
MIOpen Error: /home/atamazov/github/MLOpen2/src/solver/fft.cpp:399: Not enough workspace for FFT: expected 94740480, got 10035200
...
./bin/MIOpenDriver conv -x 5 -y 5 -W 7 -H 7 -c 832 -n 16 -k 128 -p 2 -q 2 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0 -V 1 -s 0 -w 2 -t 1 -i 6 -C /spb-hlc-fs2/1/users/atamazov/miopen-verify2
...
MIOpen(HIP): Info [FindConvFwdAlgorithm] requestAlgoCount = 2, workspace = 4076800
...
MIOpen(HIP): Info [FindSolutionImpl] FFT (not searchable)
MIOpen Error: /home/atamazov/github/MLOpen2/src/solver/fft.cpp:399: Not enough workspace for FFT: expected 161193984, got 4076800
RunForwardGPU() failed, rc = 0x7
Forward Convolution Failed: 1.79769e+308 > 1.5e-06
MIOpen(HIP): Info [FindConvBwdDataAlgorithm] requestAlgoCount = 2, workspace = 4076800
...
MIOpen(HIP): Info [FindSolutionImpl] FFT (not searchable)
MIOpen Error: /home/atamazov/github/MLOpen2/src/solver/fft.cpp:399: Not enough workspace for FFT: expected 161193984, got 4076800
...
RunBackwardGPU() failed, rc = 0x7
Backward Convolution Data Failed: 1.79769e+308 > 1.5e-06
```
</details>

Full log of failing configs: [smoke-fp32-93.sh.hip-clang.89a6cdd08.ddu_fft-invokers.0.fail.log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/5432830/smoke-fp32-93.sh.hip-clang.89a6cdd08.ddu_fft-invokers.0.fail.log.txt)

File: src/solver/fft.cpp [hunk 7]
```diff
@@ -102,34 +107,95 @@ static void cgemm_grid(size_t* global_work_size,
...
 
-    int out_n, out_c;
-    std::tie(out_n, out_c, std::ignore, std::ignore) = miopen::tien<4>(yDesc.GetLengths());
+    int temp_size = 0;
+
+    if(fwd)
+    {
+        int temp_size1 = (in_c * in_n + Padding) + (wei_k * wei_c + Padding);
+        int temp_size2 = (out_n * out_c + Padding);
+        temp_size      = temp_size1 > temp_size2 ? temp_size1 : temp_size2;
```
Comment (atamazov) [inline src/solver/fft.cpp:183]: [Question] What was wrong with `std::max`?

**Review by atamazov — CHANGES_REQUESTED** [review]
Please add logging to GWSS and I'll approve.

**Comment by atamazov** [comment]
@DrizztDoUrden Formatting issues

File: src/convolution.cpp [hunk 8]
```diff
@@ -769,6 +757,35 @@ std::size_t ConvolutionDescriptor::ForwardBackwardDataGetWorkSpaceSizeDirect(
...
+                MIOPEN_LOG_I2(sz << " < " << pair.second);
+                sz = pair.second;
+            }
+        }
+        const auto it = std::max_element(all_ws_sz.begin(),
+                                         all_ws_sz.end(),
+                                         [](auto&& l, auto&& r) { return l.second > r.second; });
+        if(it == all_ws_sz.end())
+            return 0;
+        return it->second;
```
Comment (atamazov) [inline src/convolution.cpp:780]:
```suggestion
        return sz;
```

**Review by atamazov — CHANGES_REQUESTED** [review]

**Comment by DrizztDoUrden** [comment]
Should I rename solver to "fft" to preserve solver db id?

**Comment by atamazov** [comment]
> Should I rename solver to "fft" to preserve solver db id?

Yes! We need compatibility with existing find-db.

**Comment by DrizztDoUrden** [comment]
The problem is with gemm solvers it would be impossible due to the fact there would be no single solver.

**Comment by atamazov** [comment]
[off-topic] @DrizztDoUrden 
> The problem is with gemm solvers it would be impossible due to the fact there would be no single solver.

Btw how are you going to "split" gemm? If splitting rules are trivial, then we can update names of gemm solvers directly in the perf-db and find-db.

/cc @JehandadKhan

**Comment by DrizztDoUrden** [comment]
The logic is indeed trivial. They are split by the root ifs structure from convolutionocl.cpp.

**Review by atamazov — APPROVED** [review]
LGTM. Local correctness/performance tests passed.

**Pull request merged**